primitives/src/crh/mod.rs

```diff
@@ -23,20 +23,211 @@ pub trait FixedLengthCRH {
+    /// of aggregated hash but it's actually the hash result of each of the inputs, grouped in
+    /// hash_rate chunks.
+    fn batch_evaluate_in_place(input_array: &mut[Self::Data], output_array: &mut[Self::Data]) {
+        let rate = <<Self::BaseHash as FieldBasedHash>::Parameters as FieldBasedHashParameters>::R;
+        input_array.par_chunks(rate).zip(output_array.par_iter_mut())
+            .for_each(|(inputs, output)| {
+                let mut digest = <Self::BaseHash as FieldBasedHash>::init_constant_length(rate, None);
+                inputs.iter().for_each(|input| { digest.update(input.clone()); } );
+                *output = digest.finalize().unwrap();
+            });
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use algebra::{
+        fields::mnt4753::Fr as MNT4753Fr, Field, UniformRand
+    };
+
+    use super::BatchFieldBasedHash;
+    use crate::crh::poseidon::{
+        MNT4PoseidonHash,
+        MNT4BatchPoseidonHash,
+    };
+
+    use rand_xorshift::XorShiftRng;
+    use rand::SeedableRng;
+    use crate::{FieldBasedHash, FieldBasedHashParameters};
+
+    struct DummyMNT4BatchPoseidonHash;
+
+    impl BatchFieldBasedHash for DummyMNT4BatchPoseidonHash {
+        type Data = MNT4753Fr;
+        type BaseHash = MNT4PoseidonHash;
+    }
+
+    pub(crate) fn constant_length_field_based_hash_test<H: FieldBasedHash>(
+        digest: &mut H,
+        inputs: Vec<H::Data>,
+    ) {
+        let inputs_len = inputs.len();
+
+        let final_elem = inputs[inputs_len - 1].clone();
+
+        digest.reset(None);
+        inputs.into_iter().take(inputs_len - 1).for_each(|fe| { digest.update(fe); });
+
+        // Test call to finalize() with too few inputs with respect to the declared size
+        // results in an error.
+        assert!(digest.finalize().is_err(), "Success call to finalize despite smaller number of inputs");
+
+        //Test finalize() being idempotent
+        digest.update(final_elem);
+        let output = digest.finalize().unwrap();
+        assert_eq!(output, digest.finalize().unwrap(), "Two subsequent calls to finalize gave different results");
+
+        // Test call to finalize() with too much inputs with respect to the declared size
+        // results in an error.
+        digest.update(final_elem);
+        assert!(digest.finalize().is_err(), "Success call to finalize despite bigger number of inputs");
+    }
+
+    pub(crate) fn variable_length_field_based_hash_test<H: FieldBasedHash>(
+        digest: &mut H,
+        inputs: Vec<H::Data>,
+        mod_rate: bool,
+    ) {
+        let rate = <H::Parameters as FieldBasedHashParameters>::R;
+
+        let pad_inputs = |mut inputs: Vec<H::Data>| -> Vec<H::Data> {
+            inputs.push(H::Data::one());
+            inputs.append(&mut vec![H::Data::zero(); rate - (inputs.len() % rate)]);
+            inputs
+        };
+
+        if mod_rate {
+            constant_length_field_based_hash_test(digest, inputs);
+        } else {
```

Comment:
A variable length hash function should ALWAYS add the padding.
Otherwise there are trivial collisions.

Reply:
If the hash is of variable length, but this length is always mod_rate then you cannot have collisions even if you don't add padding